tests/apollo/test_skvbc_checkpoints.py

```diff
@@ -67,3 +66,65 @@ async def test_checkpoint_creation(self, bft_network):
         checkpoint_after = await bft_network.wait_for_checkpoint(replica_id=0)
 
         self.assertEqual(checkpoint_after, 1 + checkpoint_before)
+
+    @with_trio
+    @with_bft_network(start_replica_cmd)
+    async def test_checkpoint_creation_with_f_replicas_down(self, bft_network):
+        """
+        Here we bring down a total of f replicas, verify checkpoint creation in that scenario
+
+        1) Given a BFT network, we make sure all nodes are up
+        2) Crash f replicas, excluding the current primary
+        3) Send sufficient number of client requests to trigger checkpoint protocol
+        4) Make sure checkpoint is created and agreed upon
+        """
+        bft_network.start_all_replicas()
+        skvbc = kvbc.SimpleKVBCProtocol(bft_network)
+
+        n = bft_network.config.n
+        f = bft_network.config.f
+        c = bft_network.config.c
+
+        self.assertEqual(len(bft_network.procs), n,
+                         "Make sure all replicas are up initially.")
+        current_primary = await bft_network.get_current_primary()
+
+        checkpoint_before = await bft_network.wait_for_checkpoint(replica_id=current_primary)
+
+        crashed_replicas = await self._crash_replicas_without_primary(
+            bft_network=bft_network,
+            nb_crashing=f,
+            except_replicas={current_primary}
+        )
+        self.assertFalse(current_primary in crashed_replicas)
+
+        self.assertGreaterEqual(
+            len(bft_network.procs), 2 * f + 2 * c + 1,
+            "Make sure enough replicas are up to allow a successful checkpoint creation")
+
+        await skvbc.fill_and_wait_for_checkpoint(
+            initial_nodes=bft_network.all_replicas(without=crashed_replicas),
+            checkpoint_num=1,
+            verify_checkpoint_persistency=False
+        )
+
+        checkpoint_after = await bft_network.wait_for_checkpoint(replica_id=current_primary)
+
+        self.assertEqual(checkpoint_after, 1 + checkpoint_before)
+
+    async def _crash_replicas_without_primary(
```

Comment:
Maybe we can move this to bft.py if it would be useful for other tests in the future. @teoparvanov what do you think?

Reply:
Actually @yontyon is extracting something similar in his PR (see `random_set_of_replicas`):
https://github.com/vmware/concord-bft/pull/694/files

I'd suggest to wait for him to merge and the refactor this (in a separate PR).